rust/worker/src/execution/operators/filter.rs

```diff
@@ -0,0 +1,519 @@
+    blockfile_provider: BlockfileProvider,
+    record_segment: Segment,
+    metadata_segment: Segment,
+    log_record: Chunk<LogRecord>,
+    query_ids: Option<Vec<String>>,
+    where_clause: Option<Where>,
+}
+
+impl FilterInput {
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        blockfile_provider: BlockfileProvider,
+        record_segment: Segment,
+        metadata_segment: Segment,
+        log_record: Chunk<LogRecord>,
+        query_ids: Option<Vec<String>>,
+        where_clause: Option<Where>,
+    ) -> Self {
+        Self {
+            blockfile_provider,
+            record_segment,
+            metadata_segment,
+            log_record,
+            query_ids,
+            where_clause,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct FilterOutput {
+    pub log_records: Chunk<LogRecord>,
+    pub log_oids: SignedRoaringBitmap,
+    pub compact_oids: SignedRoaringBitmap,
+}
+
+#[derive(Error, Debug)]
+pub enum FilterError {
+    #[error("Error creating record segment reader {0}")]
+    RecordSegmentReaderCreationError(#[from] RecordSegmentReaderCreationError),
+    #[error("Error materializing logs {0}")]
+    LogMaterializationError(#[from] LogMaterializerError),
+    #[error("Error filtering documents by where or where_document clauses {0}")]
+    IndexError(#[from] MetadataIndexError),
+    #[error("Error from metadata segment reader {0}")]
+    MetadataSegmentReaderError(#[from] MetadataSegmentError),
+    #[error("Error reading from record segment")]
+    RecordSegmentReaderError,
+    #[error("Invalid input")]
+    InvalidInput,
+}
+
+impl ChromaError for FilterError {
+    fn code(&self) -> ErrorCodes {
+        use FilterError::*;
+        match self {
+            RecordSegmentReaderCreationError(e) => e.code(),
+            LogMaterializationError(e) => e.code(),
+            IndexError(e) => e.code(),
+            MetadataSegmentReaderError(e) => e.code(),
+            RecordSegmentReaderError => ErrorCodes::Internal,
+            InvalidInput => ErrorCodes::InvalidArgument,
+        }
+    }
+}
+
+/// This sturct provides an abstraction over the materialized logs that is similar to the metadata segment
+pub(crate) struct MetadataLogReader<'me> {
+    // This maps metadata keys to `BTreeMap`s, which further map values to offset ids
+    // This mimics the layout in the metadata segment
+    // //TODO: Maybe a sorted vector with binary search is more lightweight and performant?
+    compact_metadata: HashMap<&'me str, BTreeMap<&'me MetadataValue, RoaringBitmap>>,
+    // This maps offset ids to documents, excluding deleted ones
+    document: HashMap<u32, &'me str>,
+    // This contains all existing offset ids that are touched by the logs
+    touched_oids: RoaringBitmap,
+    // This maps user ids to offset ids, excluding deleted ones
+    uid_to_oid: HashMap<&'me str, u32>,
+}
+
+impl<'me> MetadataLogReader<'me> {
+    pub(crate) fn new(logs: &'me Chunk<MaterializedLogRecord<'me>>) -> Self {
```

Comment:
jw - Chunk is an arc'ed type - do you need to take a reference here? Can't you take ownership over the Arc and rely on the inner lifetime?